infer/src/dotnet/ResourceLeakCSDomain.ml

```diff
@@ -46,10 +45,12 @@ module BoundsWithTop = struct
 end
 
 module ResourcesHeld = AbstractDomain.Map (AccessPath) (BoundsWithTop)
+module ResourcesTypeHeld = PrettyPrintable.MakePPMap (AccessPath)
+
 open AbstractDomain.Types
 
 (** Initializes resources to type map *)
-let type_map = ref (Hashtbl.create 100)
+let type_map = ref ResourcesTypeHeld.empty
```

Comment:
This is still mutable state, shouldn't it be part of the domain instead? Is it a global ref because it's global to the function or to the checker? If the former, either it should be made part of the domain or it should be saved and reset by [ondemand.ml with the rest of the "global state"](https://github.com/facebook/infer/blob/49ffdcea8b3a178e7fe764e9acbf4a3b0bf8242d/infer/src/backend/ondemand.ml#L79-L92) every time we change function (in particular, when analysing `foo` that calls `bar`, the type map of `foo` will leak into `bar` and be modified in the analysis of `bar` before coming back to `foo`, or is there a reason why this doesn't happen?).

Reply:
Thanks! However, I am not 100% sure what you are suggesting. I guess we can create the type mapping as an immutable state in two options: 

1. Revise 'ResourceHeld' as mapping ``AccessPath` to two values: `BoundsWithTop` and `Type`. However, I am not sure if it is feasible in ocaml?  
2. Create a separate `ResourceTypeHeld` and initialize it as a passing parameter to `Analyzer.compute_post` (https://github.com/facebook/infer/blob/39604ffab1cb76f3ef54afadebb6030a09f93b41/infer/src/dotnet/ResourceLeaksCS.ml#L184). However, this function does not take a second state. So we may need to overload this function.

Please let me know if I am on the right track.

Reply:
Either of these should be ok. You'll need to turn the type of the values of the current hash table into an abstract domain too, with `AbstractDomain.Flat` or something more fancy (*).
1. There's `AbstractDomain.Pair` to combine `BoundsWithTop` and `Type` together if a given access path always have a type whenever it is mapped to a bound and vice versa: `AbstractDomain.Map (AccessPath) (AbstractDomain.Pair (BoundsWithTop) (AbstractDomain.Flat (Typ.Name)))`
2. Or, if the maps do not necessarily have the same keys bound at all time or the bindings don't change in sync, your entire domain can be a pair eg `AbstractDomain.Pair (ResourcesHeld) (AbstactDomain.Flat (Typ.Name))`


(*) it seems that currently that type is strings, obtained from, eg, `CSharp.get_class_name`. These are typically expensive to build if that happens often, it's usually better to try and keep the original type (eg `Typ.name`) for as long as possible to avoid building those strings unless needed, hence why I  used `Typ.Name` above.

Reply:
That is very helpful! I have updated according to your suggestions. Please take a look!